File: Banker/banker.py
#############################################################
#
# Bao Pham
# Operating System
# Project #4: Banker
#
##############################################################

#function.....................................................

#print vector
def printVector (mylist):
    for val in mylist:
        print (' {:3}'.format(val), end="")
    print()
#print matrix
def printMatrice (mylist, row):
    for r in range (row):
        print ('{:3}:'.format(r), end="")
        printVector(mylist[r])
    print()
#print resource type
def printResourceType (col):
    print ("      ", end="")
    for c in range (col):
        print (' {:3}'.format(chr(ord('A')+c)), end="")
    print ()

#for taking argument
import sys
#check number of arguments
if len(sys.argv)!=2:
    print ("Error: This program can only take 1 agrument!")
    exit()

#read data from file..........................................

#open file
try:
    ifile = open(sys.argv[1], "r")
except IOError:
    print (sys.argv[1], ": no such file or directory!")
    exit()
#read all data into data
data = ifile.read().split()
p = 0
#number of processes
n = int(data[p])
p += 1
#number of resources
m = int(data[p])
p += 1
#Allocation Matrices
allocation = []
for r in range (n):
    allocation.append([int(data[p+c]) for c in range(m)])
    p += m
#Max Matrices
max = []
for r in range (n):
    max.append([int(data[p+c]) for c in range(m)])
    p += m
#Available Vector
available = []
for c in range (m):
    available.append(int(data[p+c]))
p += m
#i
i = int(data[p])
p += 2
#Request Vector
request = []
for c in range (m):
    request.append(int(data[p+c]))
#close file
ifile.close()

#Calculating and initializing variable..........................................

#Compute Need
need = []
for r in range (n):
    need.append ([max[r][c] - allocation[r][c] for c in range (m)])

#Finished Vector
finished = []
for r in range (n):
    finished.append(0)

#Work Vector
work = []
for val in available:
    work.append(val)

undone = 0
update = n

#Evaluating.....................................................................

while undone!=update:
    #update number of undone
    undone = update
    #for loop
    for r in range (n):
        #looking for qualified undone process
        if not [c for c, cj in zip(work, need[r]) if c < cj] and finished[r]==0:
            #update work
            for c in range (m): work[c] = work[c] + allocation[r][c]
            #update finished
            finished[r] = 1
            #update update
            update -= 1

#Print report...................................................................
print ("\nTHE REPORT STARTS HERE.................................")
#number of processes and resource types
print ("\nThere are ", n, " processes in the system.\n")
print ("There are ", m, " resource types.\n")
#allocation matrix
print ("The Allocation Matrix is...")
printResourceType(m)
printMatrice (allocation, n)
#max matrix
print ("The Max Matrix is...")
printResourceType(m)
printMatrice (max, n)
#need matrix
print ("The Need Matrix is...")
printResourceType(m)
printMatrice (need, n)
#available vector
print ("The Available Vector is...")
printResourceType(m)
print ("    ", end="")
printVector (available)
#state status
if undone == 0:
    print ("\nTHE SYSTEM IS IN A SAFE STATE!\n")
else:
    print ("\nTHE SYSTEM IS NOT SAFE!\n")
#request vector
print ("The Request Vector is...")
printResourceType(m)
print ('{:3}:'.format(i), end="")
printVector (request)

#evaluating the request
if not [c for c, cj in zip(need[i], request) if c < cj] and not [c for c, cj in zip(available, request) if c < cj]:
    print ("\nTHE REQUEST CAN BE GRANTED!\n")
    #update available
    for c in range (m): available[c] = available[c] - request[c]
    #reprint available vector
    print ("The Available Vector is...")
    printResourceType(m)
    print ("    ", end="")
    printVector (available)
else:
    print ("\nTHE REQUEST CANNOT BE GRANTED!")
print ("\nEND REPORT...............................................\n")

File: Banker/banker.cpp
//********************************************************************
//
// Bao Pham
// Operating Systems
// Programming Project #4: Banker
// Spring 2018
// Instructor: Dr. Dulal C. Kar
//
//********************************************************************


//*********************************************************
//
// Includes and Defines
//
//*********************************************************
#include <stdio.h>
#include <fstream>
#include "Matrix.h"

using namespace std;

//*********************************************************
//
// Function Declarations
//
//*********************************************************
// Check if in safe state
bool inSafeState (Matrix&, Matrix&, Matrix&, int);

//*********************************************************
//
// Main Function
// Takes an agrument as filename path
//
//*********************************************************
int main(int argc, char * argv[]) {
    // Number of arguments
    if (argc!=2)
    {
        printf("Error: MUST take one and only one argument as input file path\n");
        return -1;
    }

    ifstream infile; // file
    int row, col, processNum; // # of processes, # of resources, process # of request

    // Starting to read data from file................................
    infile.open(argv[1]); // Open file
    // Check error opening file
    if (infile.fail())
    {
        perror(argv[1]);
        return -1;
    }

    infile >> row; // number of processes
    infile >> col; // number of resources
    
    // nxm matrices
    Matrix allocation(row, col), max(row, col), mrequest(row, col),
        // 1xm matrices
        available(1, col), request(1, col);
    
    //Read data into matrices and vectors
    allocation.read(infile);
    max.read(infile);
    available.read(infile);
    infile >> processNum;
    infile.ignore(255, ':');
    request.read(infile);
    infile.close(); // Closing file
    Matrix need = (max - allocation);
    mrequest.setToZeroExcept(processNum, request);
    
    // Print the report.....................................
    printf("\nTHE REPORT STARTS HERE.........................\n\n");
    // The static data from input
    printf("There are %d processes in the system.\n\n", row);
    printf("There are %d resource types.\n\n", col);
    // Matrices
    printf("The Allocation Matrix is...\n");
    allocation.print(0);
    printf("\nThe Max Matrix is...\n");
    max.print(0);
    printf("\nThe Need Matrix is...\n");
    need.print(0);
    // Available vector
    printf("\nThe Available Vector is...\n");
    //allocation.printResource();
    available.print(-1);
    
    // print status of current state
    if (inSafeState(need, allocation, available, row))
        printf("\nTHE SYSTEM IS IN A SAFE STATE!\n");
    else
        printf("\nTHE SYSTEM IS NOT SAFE!\n"); 
    
    // Request Vector
    printf("\nThe Request Vector is...\n");
    //allocation.printResource();
    //printf("%*d: ", 3, processNum);
    request.print(processNum);
    
    // determine if the request is granted
    if (mrequest <= need)
    {
        
        if (request <= available)
        {
            need -= mrequest;
            available -= request;
            allocation += mrequest;
            if (inSafeState(need, allocation, available, row))
            {
                printf("3\n");
                printf("\nTHE REQUEST CAN BE GRANTED!\n");
                // calculate and print new available vector
                printf("\nThe New Available Vector is...\n");
                available.print(-1);
            }
            else 
                printf("\nTHE REQUEST CANNOT BE GRANTED!\nTHE SYSTEM IS NOT IN SAFE STATE IF REQUEST IS GRANTED!\n");
        }
        else
            printf("\nTHE REQUEST CANNOT BE GRANTED!\nTHE PROCESS MUST WAIT SINCE RESOURCES ARE NOT AVAILABLE!\n");
    }
    else 
        printf("\nTHE REQUEST CANNOT BE GRANTED!\nTHE PROCESS HAS EXCEEDED ITS MAXIMUM CLAIM!\n");
    
    printf("\nEND REPORT......................................\n\n");
    
    return 0;
}


//********************************************************************
//
// inSafeState Function
//
// Check if system in safe state
//
// Return Value
// ------------
// bool
//
// Value Parameters
// ----------------
// row      int     the number of processes
//
// Reference Parameters
// --------------------
// need             Matrix  reference to need Matrix
// allocation       Matrix  reference to allocation Matrix
// available        Matrix  reference to available Matrix
//
// Local Variables
// ---------------
// work		    Matrix		    work matrix initialize to available 
// finished     bool[]          hold finish status of processes
// last         int             number of finished processes since last update
// current      int             number of finished processes at the moment
//
//*******************************************************************
bool inSafeState(Matrix& need, Matrix& allocation, Matrix& available, int row)
{
    // Initialize variables
    Matrix work = available; // Work = Available
    bool finished[row]; // Finished = {False}
    for (int i = 0; i < row; i++)
        finished[i] = false;    
    int last = 0; // quantitizer
    int current = row; // quantitizer
    // Loop starts to determine if the current state safe
    while (last!=current)
    {
        last = current; // update undone
        for (int i = 0; i < row; i++)
        {
            // looking for undone process and has need </= work
            if (need.at(i) <= work && !finished[i])
            {
                current--; // update update
                finished[i] = true; // set process to finished
                work+=allocation.at(i); // update work
            }
        }
    }
    if (current == 0)
        return true;
    else
        return false;
}

